File: Session5/234. Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode fptr=head;
        ListNode sptr=head;
        if(head!=null)
        {
        while(fptr!=null && fptr.next!=null)
        {
            fptr=fptr.next.next;
            sptr=sptr.next;
        }
        }
        else
        {
            return true;
        }
        ListNode curr=new ListNode();
        if(fptr==null)
        {
             curr=sptr;
        }
        else
        {
             curr=sptr.next;
        }
        ListNode prev=null;
        ListNode next=null;
        while(curr!=null)
        {
          next=curr.next;
          curr.next=prev;
          prev=curr;
          curr=next;
        }
        ListNode revhead=prev;
        while(head!=null && revhead!=null)
        {
            if(head.val!=revhead.val)
            {
                return false;
            }
            head=head.next;
            revhead=revhead.next;
        }
        return true;
    }
}

File: Session5/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        fptr=head
        sptr=head
        
        while fptr is not None and fptr.next is not None:
            sptr=sptr.next
            fptr=fptr.next.next
            
        
#         // 1 2 3 4 5 6
#         // 1 2 3 4
#         // 1 3 5 null
        
#         // sptr is at next element to middle
        
#         // 1 2 3 2 1
#         // 1 2 3
#         // 1 3 1
        
#         // sptr is at middle
#         // sptr next element to start my reversal
        
        if fptr is not None:
            sptr=sptr.next
        
        curr=sptr
        prev=None
        nextv=None
        
# Reversing
        while curr is not None:
            nextv=curr.next
            curr.next=prev
            prev=curr
            curr=nextv
        p1=head  
        p2=prev
        
        # checking if it is palindrone 1st half and 2nd half
        while p2 is not None:
            if p1.val!=p2.val:
                return False
            p1=p1.next
            p2=p2.next
        
        return True